src/main/java/io/vertx/core/http/HttpServerRequest.java

```diff
@@ -478,24 +477,62 @@ default StreamPriority streamPriority() {
   /**
    * Get the cookie with the specified name.
    *
+   * NOTE: this will return just the 1st {@link Cookie} that matches the given name, to get all cookies for this name see:
+   * {@link #cookies(String)}
+   *
    * @param name  the cookie name
    * @return the cookie
    */
   default @Nullable Cookie getCookie(String name) {
-    return cookieMap().get(name);
+    for (Cookie cookie : cookies()) {
+      if (name.equals(cookie.getName())) {
+        return cookie;
+      }
+    }
+    return null;
   }
 
+  /**
+   * Get the cookie with the specified identifier.
+   *
+   * @param name  the cookie name
+   * @return the cookie
+   */
+  @Nullable Cookie getCookie(String name, String domain, String path);
+
   /**
    * @return the number of cookieMap.
    */
   default int cookieCount() {
-    return cookieMap().size();
+    return cookies().size();
   }
 
   /**
+   * @deprecated the implementation made a wrong assumption that cookies could be identified only by its name. The RFC
+   * states that the tuple of {@code <name, domain, path>} is the unique identifier.
+   *
    * @return a map of all the cookies.
    */
-  Map<String, Cookie> cookieMap();
+  @Deprecated
+  default Map<String, Cookie> cookieMap() {
+    return cookies()
+      .stream()
+      .collect(Collectors.toMap(Cookie::getName, cookie -> cookie));
+  }
+
+  /**
+   * Returns a set of parsed cookies that match the given name, or an empty set. Several cookies may share the same name
+   * but have different keys. A cookie is unique by its {@code <name, domain, path>} tuple.
+   *
+   * @param name the name to be matches
+   * @return the matching cookies or {@code null}
+   */
+  @Nullable Set<Cookie> cookies(String name);
+
+  /**
+   * @return a list with all cookies in the cookie jar.
```

Comment:
it says returns a List but it's actuallyh returning a Set

Reply:
Documentation has been updated to define the contract and tests were added to ensure it's correct.